File: public/generate-icons.js
const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const sizes = [16, 32, 64, 192, 512];
const padding = 0.15; // 15% safe padding

// Generate icons with safe padding
async function generateIcons() {
  try {
    // Load the base logo
    const sourceIcon = path.join(__dirname, 'logo.png');
    
    if (!fs.existsSync(sourceIcon)) {
      console.error(`Logo not found: ${sourceIcon}`);
      process.exit(1);
    }

    // Generate each size with padding
    for (const size of sizes) {
      const paddingPixels = Math.round(size * padding);
      const innerSize = size - (paddingPixels * 2);

      console.log(`Generating ${size}x${size} icon with ${paddingPixels}px padding...`);

      // Create icon with padding: transparent background, logo centered
      await sharp(sourceIcon)
        .resize(innerSize, innerSize, {
          fit: 'contain',
          background: { r: 0, g: 0, b: 0, alpha: 0 } // transparent
        })
        .extend({
          top: paddingPixels,
          bottom: paddingPixels,
          left: paddingPixels,
          right: paddingPixels,
          background: { r: 0, g: 0, b: 0, alpha: 0 } // transparent
        })
        .png()
        .toFile(path.join(__dirname, `favicon-${size}x${size}.png`));

      console.log(`✓ Generated favicon-${size}x${size}.png`);
    }

    // Also generate android icons
    console.log(`\nGenerating Android Chrome icons...`);
    
    // 192x192
    const padding192 = Math.round(192 * padding);
    const inner192 = 192 - (padding192 * 2);
    await sharp(sourceIcon)
      .resize(inner192, inner192, {
        fit: 'contain',
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .extend({
        top: padding192,
        bottom: padding192,
        left: padding192,
        right: padding192,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .png()
      .toFile(path.join(__dirname, `android-chrome-192x192.png`));
    console.log(`✓ Generated android-chrome-192x192.png`);

    // 512x512
    const padding512 = Math.round(512 * padding);
    const inner512 = 512 - (padding512 * 2);
    await sharp(sourceIcon)
      .resize(inner512, inner512, {
        fit: 'contain',
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .extend({
        top: padding512,
        bottom: padding512,
        left: padding512,
        right: padding512,
        background: { r: 0, g: 0, b: 0, alpha: 0 }
      })
      .png()
      .toFile(path.join(__dirname, `android-chrome-512x512.png`));
    console.log(`✓ Generated android-chrome-512x512.png`);

    // Apple touch icon
    const paddingApple = Math.round(192 * padding);
    const innerApple = 192 - (paddingApple * 2);
    await sharp(sourceIcon)
      .resize(innerApple, innerApple, {
        fit: 'contain',
        background: { r: 255, g: 255, b: 255, alpha: 1 } // white for Apple
      })
      .extend({
        top: paddingApple,
        bottom: paddingApple,
        left: paddingApple,
        right: paddingApple,
        background: { r: 255, g: 255, b: 255, alpha: 1 }
      })
      .png()
      .toFile(path.join(__dirname, `apple-touch-icon.png`));
    console.log(`✓ Generated apple-touch-icon.png`)

    console.log('\n✅ All icons generated successfully with 15% safe padding!');
    console.log('\nGenerated/Updated files:');
    sizes.forEach(size => console.log(`  - favicon-${size}x${size}.png`));
    console.log(`  - android-chrome-192x192.png`);
    console.log(`  - android-chrome-512x512.png`);
    console.log(`  - apple-touch-icon.png`);
    console.log('\nIcons now have proper padding to prevent logo overflow on mobile OS adaptive icons.');
  } catch (error) {
    console.error('Error generating icons:', error);
    process.exit(1);
  }
}

generateIcons();

File: public/generate-icons.py
#!/usr/bin/env python3
"""
Generate PWA icons with safe padding from the logo.
This ensures the logo doesn't overflow when the OS applies adaptive masking.
"""

from PIL import Image
import os

def add_padding_to_icon(source_path, output_path, size, padding_percent=15):
    """
    Add padding to an icon image.
    
    Args:
        source_path: Path to source logo image
        output_path: Path to save the padded icon
        size: Final icon size (e.g., 192, 512)
        padding_percent: Percentage of size to use as padding (default 15%)
    """
    try:
        # Open the source image
        img = Image.open(source_path).convert('RGBA')
        
        # Create a new transparent image with the target size
        new_img = Image.new('RGBA', (size, size), (255, 255, 255, 0))
        
        # Calculate padding
        padding = int(size * padding_percent / 100)
        available_size = size - (padding * 2)
        
        # Resize the logo to fit in the padded area
        img_resized = img.resize((available_size, available_size), Image.Resampling.LANCZOS)
        
        # Paste the resized logo in the center
        new_img.paste(img_resized, (padding, padding), img_resized)
        
        # Save the icon
        if output_path.endswith('.png'):
            new_img.save(output_path, 'PNG', quality=95)
        else:
            # Convert RGBA to RGB for formats that don't support transparency
            rgb_img = Image.new('RGB', (size, size), (255, 255, 255))
            rgb_img.paste(new_img, mask=new_img.split()[3])
            rgb_img.save(output_path, quality=95)
        
        print(f"✓ Generated {output_path} ({size}x{size})")
        return True
    except Exception as e:
        print(f"✗ Error generating {output_path}: {e}")
        return False

# Icon specifications
icons_to_generate = [
    ('logo.png', 'favicon-16x16.png', 16, 15),
    ('logo.png', 'favicon-32x32.png', 32, 15),
    ('logo.png', 'favicon-64x64.png', 64, 15),
    ('logo.png', 'android-chrome-192x192.png', 192, 15),
    ('logo.png', 'android-chrome-512x512.png', 512, 15),
]

# Get the public directory
public_dir = os.path.dirname(os.path.abspath(__file__))

print("Generating PWA icons with safe padding...\n")

all_success = True
for source, output, size, padding in icons_to_generate:
    source_path = os.path.join(public_dir, source)
    output_path = os.path.join(public_dir, output)
    
    if not os.path.exists(source_path):
        print(f"✗ Source file not found: {source_path}")
        all_success = False
        continue
    
    if not add_padding_to_icon(source_path, output_path, size, padding):
        all_success = False

if all_success:
    print("\n✓ All icons generated successfully with padding!")
    print("  The logos now have 15% safe padding and won't overflow on PWA installations.")
else:
    print("\n✗ Some icons failed to generate. Check the errors above.")

print("\nNext steps:")
print("1. Clear your browser cache and service worker")
print("2. Uninstall the app from your device")
print("3. Reinstall the PWA")
print("4. The icons should now display with proper padding")
